# Project 3 -- [Niharika] [Rai]

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1

In [6]:
import os 
os.listdir("/anvil/projects/tdm/data/noaa")  

['1957.csv',
 '1972.csv',
 '1854.csv',
 '1902.csv',
 '1878.csv',
 '1775.csv',
 '1808.csv',
 '1966.csv',
 '1981.csv',
 '1863.csv',
 '1911.csv',
 '1887.csv',
 '1769.csv',
 '1935.csv',
 '1784.csv',
 '1817.csv',
 '1832.csv',
 '2007.csv',
 '1990.csv',
 '1999.csv',
 '1896.csv',
 '1929.csv',
 '1944.csv',
 '1793.csv',
 '1826.csv',
 '1841.csv',
 '2016.csv',
 '1953.csv',
 '1850.csv',
 '1977.csv',
 '1859.csv',
 '1874.csv',
 '1907.csv',
 '1922.csv',
 '1771.csv',
 '1804.csv',
 '1962.csv',
 '1986.csv',
 '1868.csv',
 '1883.csv',
 '1916.csv',
 '1765.csv',
 '1931.csv',
 '1780.csv',
 '1813.csv',
 '2003.csv',
 '1789.csv',
 '1837.csv',
 '1995.csv',
 '1892.csv',
 '1925.csv',
 '1940.csv',
 '1822.csv',
 '2012.csv',
 '1949.csv',
 '1798.csv',
 '1846.csv',
 '2021.csv',
 '1958.csv',
 '1973.csv',
 '1855.csv',
 '1870.csv',
 '1903.csv',
 '1800.csv',
 '1879.csv',
 '1776.csv',
 '1809.csv',
 '1967.csv',
 '1982.csv',
 '1864.csv',
 '1912.csv',
 '1888.csv',
 '1936.csv',
 '1785.csv',
 '1818.csv',
 '1833.csv',
 '2008.csv',

In [5]:
len(os.listdir("/anvil/projects/tdm/data/noaa")) 

266

In [7]:
import pandas as pd
from pathlib import Path
Path("/anvil/projects/tdm/data/noaa/1880.csv") 

PosixPath('/anvil/projects/tdm/data/noaa/1880.csv')

In [8]:
myfiles=[]
for year in range (1880, 1884):
    file= Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file)
print(myfiles) 

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


In [9]:
myfiles = [Path(f'/anvil/projects/tdm/data/noaa/{year}.csv') for year in range (1880, 1884)] 
print(myfiles) 

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


I got this code from the 1st video, "How to find the files in a directory in Python". The link to all the videos is here: https://the-examples-book.com/programming-languages/python/looping-through-files. 
Based on part a, the files range from 1763.csv to 2023.csv. There is 1 file for 1750.csv. Then, there are other files such as status-by_year.txt, 2020_sample.csv, readme-by_year.txt and 2020_sampleB.csv. Hence, the total number of files are 266 files.

For part b, I got the code from the same video linked above. There was a typing error in part b of question 1. After watching this video, I used the correct version. 

For part c, I directly used the code given under tip. The link to that tip is here: https://the-examples-book.com/projects/current-projects/10200-2024-project03 
But I later modified that code to include the list comprehension. I took the code structure from the second video, "Example about list comprehensions" from this link: https://the-examples-book.com/programming-languages/python/looping-through-files. 

## Question 2

In [6]:
import pandas as pd
from pathlib import Path

file = Path(f'/anvil/projects/tdm/data/noaa/1880.csv')  
with open(file,"r") as f: 
    mycount = 0
    for line in f: 
        mycount += 1 
print(f'There are {mycount} records in the file called {file}') 

There are 370779 records in the file called /anvil/projects/tdm/data/noaa/1880.csv


In [7]:
for year in range(1880, 1884, 1): 
    file = Path(f'/anvil/projects/tdm/data/noaa/{year}.csv') 
    with open(file,"r") as f: 
        mycount = 0
        for line in f: 
            mycount += 1
    print(f'There are {mycount} records in the file called {file}') 

There are 370779 records in the file called /anvil/projects/tdm/data/noaa/1880.csv
There are 443270 records in the file called /anvil/projects/tdm/data/noaa/1881.csv
There are 514083 records in the file called /anvil/projects/tdm/data/noaa/1882.csv
There are 578730 records in the file called /anvil/projects/tdm/data/noaa/1883.csv


For question 2, part a and b, I followed the code structure from the video, "looping over files to check the number of records" from this link: https://the-examples-book.com/programming-languages/python/looping-through-files. However, I modified my code in context of my question. This involved using the dataset I was given for this project, using 1880

## Question 3

In [4]:
import pandas as pd
from pathlib import Path

myfiles = [] 
for year in range(1880, 1884): 
    file = Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file)
print(myfiles) 

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


In [7]:
print(myfiles[0]) 

/anvil/projects/tdm/data/noaa/1880.csv


In [8]:
myDF = [] 
myDF = pd.read_csv(myfiles[0]) 
print(myDF) 

        AGE00147705  18800101  PRCP    0 Unnamed: 4 Unnamed: 5  E  Unnamed: 7
0       AGE00147708  18800101  PRCP   36        NaN        NaN  E         NaN
1       AGE00147709  18800101  PRCP   36        NaN        NaN  E         NaN
2       AGE00147712  18800101  PRCP    0        NaN        NaN  E         NaN
3       AGE00147713  18800101  TMAX   52        NaN        NaN  E         NaN
4       AGE00147713  18800101  TMIN    0        NaN        NaN  E         NaN
...             ...       ...   ...  ...        ...        ... ..         ...
370773  USW00093852  18801231  TMAX   39        NaN        NaN  X         NaN
370774  USW00093852  18801231  TMIN  -17        NaN        NaN  X         NaN
370775  USW00094728  18801231  TMAX -122        NaN        NaN  0         NaN
370776  USW00094728  18801231  TMIN -194        NaN        NaN  0         NaN
370777  USW00094728  18801231  PRCP    0        NaN        NaN  0         NaN

[370778 rows x 8 columns]


In [9]:
myDF.columns 

Index(['AGE00147705', '18800101', 'PRCP', '0', 'Unnamed: 4', 'Unnamed: 5', 'E',
       'Unnamed: 7'],
      dtype='object')

In [11]:
myDF.head() 

,AGE00147705,18800101,PRCP,0,Unnamed: 4,Unnamed: 5,E,Unnamed: 7
0,AGE00147708,18800101,PRCP,36,NaN,NaN,E,NaN
1,AGE00147709,18800101,PRCP,36,NaN,NaN,E,NaN
2,AGE00147712,18800101,PRCP,0,NaN,NaN,E,NaN
3,AGE00147713,18800101,TMAX,52,NaN,NaN,E,NaN
4,AGE00147713,18800101,TMIN,0,NaN,NaN,E,NaN


In [12]:
myDF.tail() 

,AGE00147705,18800101,PRCP,0,Unnamed: 4,Unnamed: 5,E,Unnamed: 7
370773,USW00093852,18801231,TMAX,39,NaN,NaN,X,NaN
370774,USW00093852,18801231,TMIN,-17,NaN,NaN,X,NaN
370775,USW00094728,18801231,TMAX,-122,NaN,NaN,0,NaN
370776,USW00094728,18801231,TMIN,-194,NaN,NaN,0,NaN
370777,USW00094728,18801231,PRCP,0,NaN,NaN,0,NaN


In [14]:
header=None 
myDF.columns 

Index(['AGE00147705', '18800101', 'PRCP', '0', 'Unnamed: 4', 'Unnamed: 5', 'E',
       'Unnamed: 7'],
      dtype='object')

In [15]:
pd.read_csv(myfiles[0],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"]) 

,id,date,element_code,value,mflag,qflag,sflag,obstime
0,AGE00147705,18800101,PRCP,0,NaN,NaN,E,NaN
1,AGE00147708,18800101,PRCP,36,NaN,NaN,E,NaN
2,AGE00147709,18800101,PRCP,36,NaN,NaN,E,NaN
3,AGE00147712,18800101,PRCP,0,NaN,NaN,E,NaN
4,AGE00147713,18800101,TMAX,52,NaN,NaN,E,NaN
...,...,...,...,...,...,...,...,...
370774,USW00093852,18801231,TMAX,39,NaN,NaN,X,NaN
370775,USW00093852,18801231,TMIN,-17,NaN,NaN,X,NaN
370776,USW00094728,18801231,TMAX,-122,NaN,NaN,0,NaN
370777,USW00094728,18801231,TMIN,-194,NaN,NaN,0,NaN


In [18]:
mydataframes = [] 
for i in range(0,4): 
    mydataframes.append(pd.read_csv(myfiles[i],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"])) 
print(mydataframes) 

[                 id      date element_code  value mflag qflag sflag  obstime
0       AGE00147705  18800101         PRCP      0   NaN   NaN     E      NaN
1       AGE00147708  18800101         PRCP     36   NaN   NaN     E      NaN
2       AGE00147709  18800101         PRCP     36   NaN   NaN     E      NaN
3       AGE00147712  18800101         PRCP      0   NaN   NaN     E      NaN
4       AGE00147713  18800101         TMAX     52   NaN   NaN     E      NaN
...             ...       ...          ...    ...   ...   ...   ...      ...
370774  USW00093852  18801231         TMAX     39   NaN   NaN     X      NaN
370775  USW00093852  18801231         TMIN    -17   NaN   NaN     X      NaN
370776  USW00094728  18801231         TMAX   -122   NaN   NaN     0      NaN
370777  USW00094728  18801231         TMIN   -194   NaN   NaN     0      NaN
370778  USW00094728  18801231         PRCP      0   NaN   NaN     0      NaN

[370779 rows x 8 columns],                  id      date element_code  val

In [30]:
for i in range(0, 4): 
    print(mydataframes[i].columns) 

Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')


Something unexpected that I saw in part a was that the last 4 columns are unnamed. I got the code structure from question 1, tip of project 3. To display the column names, head and tail, I types those fuctions by having memorized them from the works I did in the previous 2 projects. I also referred to part d which said to use myDF.columns to display the column names. I also used this website as a reference: https://note.nkmk.me/en/python-pandas-len-shape-size/. For part b, I used the hint: header=None and it gave me the same result that there are no column names. Instead of keeping everything on the same line, I created my own code structure and wrote them out separately in different lines. For part c, I used the code structure from the question itself. For part d, I got the code structure from the video, "reading in data frames in a loop" from this link: https://the-examples-book.com/programming-languages/python/looping-through-files. However, I added an additional step of print(mydataframes) so that I can see all the rows and columns at the same time. Furthermore, I even checked my work using loop to see that each of the data files from 1880.csv to 1883.csv have had a change in column names. I got the code structure from a video called, "". linked above. But, instead of using shape, I used columns by taking an exemplar reference of myDF.columns from part d of the question. 

## Question 4

In [11]:
import pandas as pd
from pathlib import Path

myfiles = [] 
for year in range(1880, 1884): 
    file = Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file) 
    
myDF = [] 
myDF = pd.read_csv(myfiles[0]) 
print(myDF) 

mydataframes = [] 
for i in range(0,4): 
    mydataframes.append(pd.read_csv(myfiles[i],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"])) 
    
for myDF in mydataframes: 
    print(myDF['element_code'].unique()) 

        AGE00147705  18800101  PRCP    0 Unnamed: 4 Unnamed: 5  E  Unnamed: 7
0       AGE00147708  18800101  PRCP   36        NaN        NaN  E         NaN
1       AGE00147709  18800101  PRCP   36        NaN        NaN  E         NaN
2       AGE00147712  18800101  PRCP    0        NaN        NaN  E         NaN
3       AGE00147713  18800101  TMAX   52        NaN        NaN  E         NaN
4       AGE00147713  18800101  TMIN    0        NaN        NaN  E         NaN
...             ...       ...   ...  ...        ...        ... ..         ...
370773  USW00093852  18801231  TMAX   39        NaN        NaN  X         NaN
370774  USW00093852  18801231  TMIN  -17        NaN        NaN  X         NaN
370775  USW00094728  18801231  TMAX -122        NaN        NaN  0         NaN
370776  USW00094728  18801231  TMIN -194        NaN        NaN  0         NaN
370777  USW00094728  18801231  PRCP    0        NaN        NaN  0         NaN

[370778 rows x 8 columns]
['PRCP' 'TMAX' 'TMIN' 'SNOW' 'SNWD' '

In [13]:
for myDF in mydataframes: 
    print(myDF['element_code'].value_counts()['SNOW']) 

33700
36451
40648
53417


For part a, I used the code structure from the video, "finding unique elements in a column of a data frame", linked in the examples book: https://the-examples-book.com/programming-languages/python/looping-through-files. But before that, I also had to input part of the codes that I wrote in question 3 because otherwise it was giving me an error as Path, pd, mydataframes, and myfiles were not defined. To ensure that the element_codes are different, I used the len fuction to calculate the number of elements in each of these dataframes. Hence, the different values indicate that each data frame had a different information displayed under the element_code column. 

For part b, I got the code structure from a video titled, "finding value counts in columns of a list of data frames" linked in the examples book: https://the-examples-book.com/programming-languages/python/looping-through-files. 

## Question 5

In [3]:
import pandas as pd
from pathlib import Path
myfiles=[]
for year in range (1880, 1884):
    file= Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file) 
    
count = 0
for file in myfiles:
    for myDF in pd.read_csv(file,names=["id","date","element_code","value","mflag","qflag","sflag","obstime"],chunksize =10000):
        count += len(myDF[myDF['element_code'] == 'SNOW'])

print(count) 

164216


In [4]:
count = 0
for file in myfiles:
    for myDF in pd.read_csv(file,names=["id","date","element_code","value","mflag","qflag","sflag","obstime"],chunksize =10000):
        for index, row in myDF.iterrows():
            if row['element_code'] == 'SNOW':
                count += 1

print(count) 

164216


For part a, I got all the codes from the question. Both versions of the program lead to the same answer. This answer is found by adding: 33700+36451+40648+53417 = 164216. This is because the question asked to find how many times 'SNOW' occured in each of the data frames and hence, the answer was the sum of the values in each of the 4 data frames where 'SNOW' occurred. According to this website: https://www.geeksforgeeks.org/how-to-load-a-massive-file-as-small-chunks-in-pandas/, chunksize is basicially the number of rows that are read at 1 given time by the file in pandas. So, if the chunksize in this case is 10000 and the number of rows in the 1880.csv file is 370779, then the number of rows read at once are in 370779/10000 = 37 chunks. The remaining 0779 form the 38th chunk. However, another important point I found out from the same website was that chunksize does not affect the columns in anyway. So, the 8 columns that we have in each of the data frames from 1880.csv to 1883.csv stay as they are. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.